### PR TITLE
Exposing Windows::ApplicationModel::AppService APIs in python

### PR DESCRIPTION
**Pull request by alanch-ms**

### CODE REVIEW

**Comment by alanch-ms** [comment]
@alecont

**Comment by alanch-ms** [comment]
@munyirik Please CR :)

**Comment by munyirik** [comment]
Will CR today...

**Comment by munyirik** [comment]
cc. @bfjelds @jessekaplan @trmck

**Comment by munyirik** [comment]
@alanch-ms could you share a sample/snippet of usage? (sorry late start:-))

**Comment by alanch-ms** [comment]
Added you on a mail thread with sample code of the usage.

Sent from Outlookhttp://aka.ms/weboutlook

File: extensions/winrt/_app_service/AppServiceClosedEventArgs.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        { "_CreateFromIInspectable", WinRTHelper<thisType>::CreateFromInspectable, METH_CLASS | METH_VARARGS, NULL },
+        { NULL, NULL },
+    };
+
+#pragma endregion methods
+
+    static const char type_name[] = "WinRT.ApplicationModel.AppService.AppServiceClosedEventArgs";
+    static const char docstring[] = "Provides data for the AppServiceConnection.ServiceClosed event that occurs when the other endpoint closes connection to the app service.";
+
+    PyTypeObject Type = {
```
Comment (munyirik) [inline extensions/winrt/_app_service/AppServiceClosedEventArgs.cpp:44]: Can you define PyTypeObject in one file?

File: extensions/winrt/_app_service/AppServiceConnection.cpp [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+            }
+
+            if (!self->dict)
+            {
+                self->dict = PyDict_New();
+            }
+
+            auto thisObj = ref new AppServiceConnection();
+
+            using namespace Windows::Foundation;
```
Comment (munyirik) [inline extensions/winrt/_app_service/AppServiceConnection.cpp]: is this using intentional?

Comment (munyirik) [inline extensions/winrt/_app_service/AppServiceConnection.cpp]: is this using intentional?

File: extensions/winrt/_app_service/AppServiceClosedEventArgs.cpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+{
+    typedef AppServiceClosedEventArgs thisType;
+#pragma region getset
+
+    static PyObject * getStatus(PyObject *obj, void *closure)
+    {
+        PyObject* const FailureValue = NULL;
+        return ExceptionGuard([=]
+        {
+            PyWinRTObject *self = reinterpret_cast<PyWinRTObject*>(obj);
```
Comment (munyirik) [inline extensions/winrt/_app_service/AppServiceClosedEventArgs.cpp:18]: where is PyWinRTObject defined?

File: extensions/winrt/_winrt/include/ExceptionGuard.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    }
+    catch (Platform::Exception ^ e)
+    {
+        // assert that python exception is not already set
+        assert(!PyErr_Occurred());
+
+        auto msg = e->Message;
+        auto pyMsg = PyUnicode_FromPlatformString(msg);
+        PyErr_SetObject(PyExc_RuntimeError, pyMsg.Get());
+        return failureValue;
```
Comment (munyirik) [inline extensions/winrt/_winrt/include/ExceptionGuard.h:81]: maybe return e->HResult when failure happens instead?

File: extensions/winrt/_app_service/AppServiceConnection.cpp (hunk 2 above)
Comment (alanch-ms) [inline extensions/winrt/_app_service/AppServiceConnection.cpp]: No.  I moved it to the top of the file, but forgot to remove this one.  I'll remove it. Thanks.

File: extensions/winrt/_winrt/include/ExceptionGuard.h (hunk 4 above)
Comment (alanch-ms) [inline extensions/winrt/_winrt/include/ExceptionGuard.h:81]: You can't return e->HResult.  Python doesn't know how to handle it.  AFAIK, Python has 2 conventions: return PyObject_, and return Int.   For PyObject_, return null means failure and any thing else means success.  For int, return 0 on success and anything else means failure.

File: extensions/winrt/_app_service/AppServiceClosedEventArgs.cpp (hunk 3 above)
Comment (alanch-ms) [inline extensions/winrt/_app_service/AppServiceClosedEventArgs.cpp:18]: In extensions\winrt_winrt\include\winrt_object.h

File: extensions/winrt/_app_service/AppServiceClosedEventArgs.cpp (hunk 1 above)
Comment (alanch-ms) [inline extensions/winrt/_app_service/AppServiceClosedEventArgs.cpp:44]: No.  You need a different instance of PyTypeObject for each python type that you are defining.  See https://docs.python.org/3/extending/newtypes.html.

**Comment by munyirik** [comment]
:+1:

**Comment by alecont** [comment]
looks good

**Pull request merged**